File: problems/0045-jump-game-ii/solution.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        /*
        0045. Jump Game II

        Given an array nums where each element represents the maximum jump length from that position,
        return the minimum number of jumps required to reach the last index.

        Approach:
        We use a greedy strategy:
        - We track the current range (window) that can be reached with the current number of jumps.
        - We also track the farthest index we can reach during this window.
        - Whenever we reach the end of the current window, we increment the jump counter
          and update the window to the farthest index found so far.

        Time complexity: O(n), where n is the length of nums.
        We make a single pass through the array.

        Space complexity: O(1), we only use a few integer variables.
        */

        int n = nums.size();
        int jumps = 0;         // Number of jumps made so far.
        int farthest = 0;      // The farthest index reachable in current scan.
        int end = 0;           // End of the current jump's window.

        for (int i = 0; i < n - 1; ++i) {
            // Update the farthest index we can reach from index i.
            farthest = max(farthest, i + nums[i]);

            // If we reached the end of the current jump's range.
            if (i == end) {
                ++jumps;          // Make a jump.
                end = farthest;   // Move window to the farthest reach.
            }
        }

        return jumps;
    }
};

File: problems/0045-jump-game-ii/solution.py
from typing import List

class Solution:
    def jump(self, nums: List[int]) -> int:
        """
        0045. Jump Game II

        Given an array nums where each element represents the maximum jump length from that position,
        return the minimum number of jumps required to reach the last index.

        Approach:
        Use a greedy strategy:
        - Track the current window (range of indices reachable with the current number of jumps).
        - Track the farthest index reachable from the current window.
        - When we reach the end of the current window, increment jump count and move to the farthest.

        Time complexity: O(n), where n is the length of nums.
        Each element is visited once to update the farthest reachable index.

        Space complexity: O(1), only a few integer variables are used.
        """

        n = len(nums)
        jumps = 0              # Number of jumps made so far.
        farthest = 0           # The farthest index reachable in the current scan.
        end = 0                # End of the current jump's window.

        for i in range(n - 1):
            # Update the farthest index reachable from this position.
            farthest = max(farthest, i + nums[i])

            # If we reach the end of the current jump window:
            if i == end:
                jumps += 1       # Make a jump.
                end = farthest   # Update window to farthest reach.

        return jumps

